File: Data Structures & Algorithms/longest-consecutive-sequence/submission-0.java
class Solution {
    public int longestConsecutive(int[] nums) {
        //Map<Integer, List<Integer>> map = new HashMap<Integer, List<Integer>>();
        //首先找到可以作为头的数字，注意不能on^2的方式，一个数对比一圈，而是用set的特性，分为两步得到这样的数字
        //得到以后，对每个数字进行数组长度的累积，甚至不用存下来这个数组
        Set<Integer> set = new HashSet<>();
        for (int num: nums){
            set.add(num);
        }
        int longest = 0;
        for (int num: nums){
            if(!set.contains(num-1)){//如果不存在就可以是头，因为不是每个都是头，所以复杂度并非n^2，最坏
                int len =  1;
                while(set.contains(num + len)){
                    len++;
                }
                longest = Math.max(len, longest);
            }
        }
        return longest;

    }
}

File: Data Structures & Algorithms/house-robber-ii/submission-0.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        length = len(nums)
        if length == 1:return nums[0]

        def dp(x):
            length = len(x)
            if length == 1:return x[0]
            dic = {}
            dic[0] = x[0]
            dic[1] = max(x[0], x[1])
            for i in range(2, length):
                dic[i] = max(dic[i-2]+x[i], dic[i-1])
            return dic[length-1]

        return max(dp(nums[:length-1]), dp(nums[1:length]))

        

        
        
        



